packages/tpetra/core/src/Tpetra_Details_unpackCrsMatrixAndCombine_def.hpp

```diff
@@ -215,190 +212,204 @@ struct UnpackCrsMatrixAndCombineFunctor {
+        (int) lid_no, (int) num_entries_in_row, (int) num_ent_out
+      );
+      Kokkos::atomic_compare_exchange_strong(error_code.data(), 0, 23);
     }
 
-    // Column indices come in as global indices, in case the
-    // source object's column Map differs from the target object's
-    // (this's) column Map, and must be converted local index values
-    for (size_t k = 0; k < num_ent; ++k) {
-      lids_out(k) = local_col_map.getLocalElement (gids_out(k));
-    }
+    constexpr bool matrix_has_sorted_rows = true; // see #6282
+    Kokkos::parallel_for(
+      Kokkos::TeamThreadRange(team_member, num_entries_in_batch),
+      KOKKOS_LAMBDA(const LO& j)
+      {
+        size_t distance = 0;
+
+        GO gid_out;
+        distance = j * bytes_per_gid;
+        (void) PackTraits<GO>::unpackValue(gid_out, gids_in + distance);
+        auto lid_out = local_col_map.getLocalElement(gid_out);
+
+        // Column indices come in as global indices, in case the
+        // source object's column Map differs from the target object's
+        // (this's) column Map, and must be converted local index values
+
+        // assume that ST is default constructible
+        ST val_out;
+        distance = j * bytes_per_value;
+        (void) PackTraits<ST>::unpackValue(val_out, vals_in + distance);
+
+        if (combine_mode == ADD) {
+          // NOTE (mfh 20 Nov 2019) Must assume atomic is required, unless
+          // different threads don't touch the same row (i.e., no
+          // duplicates in incoming LIDs list).
+          const bool use_atomic_updates = atomic;
+          (void)local_matrix.sumIntoValues(
+            import_lid,
+            &lid_out,
+            1,
+            &val_out,
+            matrix_has_sorted_rows,
+            use_atomic_updates
+          );
+        } else if (combine_mode == REPLACE) {
+          // NOTE (mfh 20 Nov 2019): It's never correct to use REPLACE
+          // combine mode with multiple incoming rows that touch the same
+          // target matrix entries, so we never need atomic updates.
+          const bool use_atomic_updates = false;
+          (void)local_matrix.replaceValues(
+            import_lid,
+            &lid_out,
+            1,
+            &val_out,
+            matrix_has_sorted_rows,
+            use_atomic_updates
+          );
+        } else {
+          // should never get here
+          printf(
+            "*** Error: UnpackCrsMatrixAndCombineFunctor: "
+            "At row %d, an unknown error occurred during unpack\n", (int) lid_no
+          );
+          Kokkos::atomic_compare_exchange_strong(error_code.data(), 0, 31);
+        }
+      }
+    );
 
-    // Combine the values according to the combine_mode
-    const LO* const lids_raw = const_cast<const LO*> (lids_out.data ());
-    const ST* const vals_raw = const_cast<const ST*> (vals_out.data ());
-    LO num_modified = 0;
+    team_member.team_barrier();
```

Comment:
The error code records the first error. Thus, after team barrier, you might want to check the error code and return immediately (another thread may already report an error). After this kernels is finished with a non-zero error code, a user can find the first problem without waiting for finishing this kernel.   

and result in early termination o

Reply:
I am not sure how much checking the error code every time costs. Maybe checking the error code every time does not make sense. When error happens, who care if the code tells you that "you are wrong" earlier or not. I think that it can go as it is.